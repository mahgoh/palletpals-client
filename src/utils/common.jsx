export function classNames(...classes) {
  return classes.filter(Boolean).join(' ')
}

export function productImageURL(imageId) {
  return `${GLOBAL.API_URL}/product-images/${imageId}`
}

export function parseDateTime(dateTimeString, locale = 'en') {
  const datetime = new Date(dateTimeString)
  const options = {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: 'numeric',
    minute: '2-digit',
  }

  return datetime.toLocaleString(locale, options)
}

export function formatPrice(price) {
  return `CHF ${price.toFixed(2)}`
}

export function formatOrderId(id) {
  return id.toString().padStart(6, '0')
}
