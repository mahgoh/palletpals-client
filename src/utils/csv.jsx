export async function processCSV(file) {
  try {
    const csv = await readCSV(file)
    const parsed = parseCSV(csv)

    return parsed
  } catch (e) {
    throw e
  }
}

function readCSV(file) {
  return new Promise((resolve, reject) => {
    if (file === null) reject('No file selected')

    // Check if the file is an image.
    if (file.type && !file.type === 'text/csv') {
      console.log('File is not csv.', file.type, file)
      reject('File is not csv.')
    }

    const reader = new FileReader()
    reader.addEventListener('load', (event) => {
      resolve(event.target.result)
    })
    reader.readAsText(file)
  })
}

/**
 *
 * Target:
 * {
 *   "km": [10,20,30],
 *   "pallets": [1,2,3],
 *   "price": [[5,6,7],[9,11,13],[15,17,19]]
 * }
 * @param {string} csv
 * @returns
 */

function parseCSV(csv) {
  const lines = csv.split('\n')

  // how entries are separated
  const delimiter = csv.includes(',') ? ',' : ';'

  const head = lines.shift()

  // palletArray is the first row except the first column
  // integers
  const palletArray = head
    .split(delimiter)
    .slice(1)
    .map((p) => parseInt(p))

  // kmArray is the first column except the first row
  // integers
  const kmArray = lines.map(
    (line) => line.split(delimiter).map((km) => parseInt(km))[0]
  )

  // priceMatrix is the rest of the table as two dimensional array
  // floats
  const priceMatrix = lines.map((line) =>
    line
      .split(delimiter)
      .slice(1)
      .map((p) => parseFloat(p))
  )

  return {
    kmArray,
    palletArray,
    priceMatrix,
  }
}
