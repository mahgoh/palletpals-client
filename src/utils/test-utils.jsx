import { cloneElement } from 'react'
import { BrowserRouter } from 'react-router-dom'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'
import i18n from '../i18n'

import { AuthProvider } from '@/services/auth'
import { AppearanceProvider } from '@/services/appearance'
import { CartProvider } from '@/services/cart'
import { LanguageProvider } from '@/services/language'
import { NotificationProvider } from '@/services/notification'

/**
 * Render component with router wrapper and option for specifying the current route.
 *
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouter = (component, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)

  return {
    ...render(component, { wrapper: BrowserRouter }),
  }
}

/**
 * Render component with router wrapper and i18next provider.
 * @param {JSX} component
 * @param {Object} options
 * @returns
 */
const renderWithRouterWithi18n = (component, options) => {
  const comp = cloneElement(component, {
    changeLanguage: (lng) => {
      i18n.changeLanguage(lng)
      rerender(<I18nextProvider i18n={i18n}>{comp}</I18nextProvider>)
    },
  })
  const defaultRender = renderWithRouter(
    <I18nextProvider i18n={i18n}>{comp}</I18nextProvider>,
    options
  )
  const { rerender } = defaultRender
  return defaultRender
}

const renderComplete = (component, options) => {
  return {
    ...renderWithRouterWithi18n(
      <AuthProvider>
        <AppearanceProvider>
          <LanguageProvider>
            <CartProvider>
              <NotificationProvider>{component}</NotificationProvider>
            </CartProvider>
          </LanguageProvider>
        </AppearanceProvider>
      </AuthProvider>,
      options
    ),
  }
}

/**
 * Render component with specified renderer
 *
 * @param {JSX} component
 * @param {Object} param1
 * @returns
 */
const customRender = (component, { renderer = 'default', ...args } = {}) => {
  switch (renderer) {
    case 'router':
      return renderWithRouter(component, { ...args })
    case 'router-i18n':
      return renderWithRouterWithi18n(component, { ...args })
    case 'complete':
      return renderComplete(component, { ...args })
    case 'default':
    default:
      return {
        ...render(component),
      }
  }
}

export * from '@testing-library/react'
// export { default as userEvent } from '@testing-library/user-event'
export { customRender as render }
