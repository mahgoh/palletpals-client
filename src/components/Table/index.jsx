import PropTypes from 'prop-types'
import { classNames, formatPrice } from '@/utils/common'

export function FinancialTable({ rows }) {
  return (
    <table className="min-w-full divide-y divide-gray-200 dark:divide-gray-700">
      <tbody>
        {rows.map((row, i) => (
          <tr
            key={i}
            className="border-b border-gray-200 last:border-b-0 dark:border-gray-700"
          >
            <td className="py-4 pl-4 pr-3 text-sm sm:pl-6 md:pl-0">
              <div className="font-medium text-gray-900 dark:text-gray-100">
                {row.key}
              </div>
            </td>
            <td
              className={classNames(
                'py-4 pl-3 pr-4 text-right text-sm sm:pr-6 md:pr-0',
                row.bold
                  ? 'font-semibold text-gray-900 dark:text-gray-100'
                  : 'text-gray-500 dark:text-gray-300'
              )}
            >
              {formatPrice(row.value)}
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default function Table({ rows }) {
  if (!rows || rows.length === 0) {
    return null
  }

  return (
    <div className="mt-4 flex w-full flex-col">
      <div className="inline-block min-w-full py-2 align-middle">
        <div className="overflow-hidden ring-1 ring-gray-200 dark:ring-gray-700">
          <table className="min-w-full ">
            <tbody className="divide-y divide-gray-200 dark:divide-gray-700">
              {rows.map((row, i) => (
                <tr
                  key={i}
                  className="divide-x divide-gray-200 dark:divide-gray-700"
                >
                  <td className="whitespace-nowrap py-4 pl-4 pr-4 text-sm font-medium text-gray-900 dark:text-gray-100 sm:pl-6">
                    {row.key}
                  </td>
                  <td className="whitespace-nowrap py-4 pl-4 pr-4 text-sm text-gray-500 dark:text-gray-300  sm:pr-6">
                    {row.value}
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  )
}

Table.propTypes = {
  rows: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    })
  ).isRequired,
}
