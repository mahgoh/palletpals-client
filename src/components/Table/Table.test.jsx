import React from 'react'
import { render, screen } from '@/utils/test-utils'
import Table from '.'

describe('table', () => {
  it('should render', () => {
    const rows = [
      {
        key: 'First',
        value: 10,
      },
      {
        key: 'Second',
        value: 'Something',
      },
    ]

    render(<Table rows={rows} />)
    expect(screen.getByRole('table')).toBeTruthy()
    expect(screen.getByRole('table')).toMatchSnapshot()
  })

  it('should render with no rows', () => {
    render(<Table rows={[]} />)
    expect(screen.queryByRole('table')).toBeNull()
    expect(screen.queryByRole('table')).toMatchSnapshot()
  })

  it('should not render children', () => {
    render(
      <Table rows={[]}>
        <div>should not be rendered</div>
      </Table>
    )
    expect(screen.queryByRole('table')).toBeNull()
    expect(screen.queryByRole('table')).toMatchSnapshot()
  })

  it('should not render', () => {
    const { container } = render(<Table />)
    expect(container).toBeEmptyDOMElement()
  })
})
