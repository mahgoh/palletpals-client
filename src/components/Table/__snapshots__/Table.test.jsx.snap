// Vitest Snapshot v1

exports[`table > should not render children 1`] = `null`;

exports[`table > should render 1`] = `
<table
  class="min-w-full "
>
  <tbody
    class="divide-y divide-gray-200 dark:divide-gray-700"
  >
    <tr
      class="divide-x divide-gray-200 dark:divide-gray-700"
    >
      <td
        class="whitespace-nowrap py-4 pl-4 pr-4 text-sm font-medium text-gray-900 dark:text-gray-100 sm:pl-6"
      >
        First
      </td>
      <td
        class="whitespace-nowrap py-4 pl-4 pr-4 text-sm text-gray-500 dark:text-gray-300  sm:pr-6"
      >
        10
      </td>
    </tr>
    <tr
      class="divide-x divide-gray-200 dark:divide-gray-700"
    >
      <td
        class="whitespace-nowrap py-4 pl-4 pr-4 text-sm font-medium text-gray-900 dark:text-gray-100 sm:pl-6"
      >
        Second
      </td>
      <td
        class="whitespace-nowrap py-4 pl-4 pr-4 text-sm text-gray-500 dark:text-gray-300  sm:pr-6"
      >
        Something
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`table > should render with no rows 1`] = `null`;
